atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransaction.java

```diff
@@ -442,13 +447,49 @@ public void put(TableReference tableRef, Map<Cell, byte[]> values) {
         super.put(tableRef, values);
     }
 
+    // for now, ignores negative lookups because this method is for metrics and that doesn't rely on it
+    private Set<TableReference> someTablesReadExcludingWrites() {
+        Set<TableReference> tablesRead = new HashSet<>();
+        tablesRead.addAll(Maps.filterEntries(readsByTable,
+                entry -> !writesByTable.containsKey(entry.getKey())
+                        || !writesByTable.get(entry.getKey()).keySet().containsAll(entry.getValue().keySet()))
+                .keySet());
+        return tablesRead;
+    }
+
+    private static String tableConflictMetricName(String name) {
+        return MetricRegistry.name(SerializableTransaction.class, "tableConflictHandling", name);
+    }
+
     @Override
     protected void throwIfReadWriteConflictForSerializable(long commitTimestamp) {
-        Transaction ro = getReadOnlyTransaction(commitTimestamp);
-        verifyRanges(ro);
-        verifyColumnRanges(ro);
-        verifyCells(ro);
-        verifyRows(ro);
+        TableTransactionConflictManager.Status status =
+                tableConflictsTransaction.commit(
+                        commitTimestamp,
+                        writesByTable.keySet(),
+                        someTablesReadExcludingWrites());
+        if (status == TableTransactionConflictManager.Status.ERROR) {
+            metricsManager.getRegistry().meter(tableConflictMetricName("error")).mark();
+        }
+        try {
+            Transaction ro = getReadOnlyTransaction(commitTimestamp);
+            verifyRanges(ro);
+            verifyColumnRanges(ro);
+            verifyCells(ro);
+            verifyRows(ro);
+        } catch (TransactionSerializableConflictException e) {
+            if (status == TableTransactionConflictManager.Status.NO_READ_WRITE_CONFLICTS) {
+                metricsManager.getRegistry().meter(tableConflictMetricName("unnecessaryConflicts")).mark();
+            }
+            throw e;
+        }
+        if (status == TableTransactionConflictManager.Status.NO_READ_WRITE_CONFLICTS) {
+            metricsManager.getRegistry().meter(tableConflictMetricName("couldBeAvoided.txns")).mark();
+            metricsManager.getRegistry().meter(tableConflictMetricName("couldBeAvoided.cellLoads"))
+                    .mark(cellsRead.values().stream().mapToInt(Collection::size).asLongStream().sum());
+        } else if (status == TableTransactionConflictManager.Status.CANNOT_ADVISE) {
+            metricsManager.getRegistry().meter(tableConflictMetricName("couldNotAdvise")).mark();
+        }
```

Comment:
Might be missing something, but I'm a bit concerned about this case. It could be okay if it won't impact the metrics excessively - it's also very possible the situation I describe is impossible because of something I missed.

Suppose there are two transactions `A` and `B`; `A` has a start timestamp of 10 and commit timestamp of 20, while B has a start timestamp of 11 and commit timestamp of 19. `A` and `B` read from and write to some cells in the table `foo`. Now `A` reaches the `TableTransactionConflictManager` first; since no transactions are running, it has `NO_READ_WRITE_CONFLICTS` and then the thread for `A` goes to sleep. `B` then reaches that, and registers a write-conflict on `A`, and then finds it has a conflicting read with `A`, so we `CANNOT_ADVISE` for `B`. Now `B` does the check and commits normally. `A` then wakes up and tries to replay its read at 20, which fails, thus we claim that we have one `unnecessaryConflict` when it seems like the transactions have to conflict in some way.

The graph algorithms in the Fekete paper seem to be written with foreknowledge of what transactions are possible. There is probably some clever logic that can be applied to make these work in an online context (probably we need to register our read- or write- sets earlier in the AtlasDB protocol), though it's not immediately apparent to me.